### ISSUE TITLE
SpaceAroundBlockParameters, MultilineBlockLayout, with Multi-Line Block Parameters

### ISSUE THREAD

**Issue by Ajedi32** [rubocop-hq/rubocop#3359]
I'm using Rubocop in my project, and I'm running into a rather onerous situation with a few of the Rubocop warnings while trying to define an RSpec shared example with named parameters.

I have the following code, which Rubocop fails on `Style/SpaceAroundBlockParameters`, and `Style/MultilineBlockLayout`:

``` ruby
RSpec.shared_examples 'config files work as expected' do |
  config_home_param: :config_home,
  service_name_param: :service_name,
  config_dirs_param: :config_dirs|
  # ...
end
```

``` text
spec/helpers/shared_config_file_examples.rb:10:58: C: Block argument expression is not on the same line as the block start.
RSpec.shared_examples 'config files work as expected' do | ...
                                                         ^
spec/helpers/shared_config_file_examples.rb:10:59: C: Space before first block parameter detected.
RSpec.shared_examples 'config files work as expected' do | ...

spec/helpers/shared_config_file_examples.rb:11:35: C: Extra space before block parameter detected.
  config_home_param: :config_home, ...

spec/helpers/shared_config_file_examples.rb:12:37: C: Extra space before block parameter detected.
  service_name_param: :service_name, ...
```

I can't put the parameters on one line because that will fail `Metrics/LineLength`:

``` ruby
RSpec.shared_examples 'config files work as expected' do |config_home_param: :config_home, service_name_param: :service_name, config_dirs_param: :config_dirs|
  # ...
end
```

``` text
spec/helpers/shared_config_file_examples.rb:10:81: C: Line is too long. [158/80]
RSpec.shared_examples 'config files work as expected' do |config_home_param: :config_home, service_name_param: :service_name, config_dirs_param: :config_dirs|
                                                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

I also can't move the parameter definitions to the block body because that fails `Style/OptionHash`:

``` ruby
RSpec.shared_examples 'config files work as expected' do |options = {}|
  backup_suffix_param = options[:backup_suffix_param] || :backup_suffix
  config_home_param   = options[:config_home_param]   || :config_home
  service_name_param  = options[:service_name_param]  || :service_name
  # ...
end
```

``` text
spec/helpers/shared_config_file_examples.rb:10:59: C: Prefer keyword arguments to options hashes.
RSpec.shared_examples 'config files work as expected' do |options = {}|
                                                          ^^^^^^^^^^^^
```

Importantly, **I don't want to _completely_ disable any of these restrictions**, because in the case of single-line block parameters I actually agree with them.

Could `SpaceAroundBlockParameters` and `MultilineBlockLayout` be changed to work better in the case of multi-line block parameters, or at least have an option to turn them off in that specific case (and only that case)? Or is there some other prefered style that I'm missing here which I could use?
## RuboCop version

```
$ rubocop -V
0.41.2 (using Parser 2.3.1.2, running on ruby 2.3.0 x86_64-linux)
```

**Comment by jonas054** [comment]
This is my suggestion for how to format the above code:
```ruby
RSpec.shared_examples 'config files work as expected' \
  do |config_home_param: :config_home,
      service_name_param: :service_name,
      config_dirs_param: :config_dirs|
  # ...
end
```
This gives us the following reported offenses:
```
/tmp/3359.rb:2:6: C: Style/MultilineBlockLayout: Block argument expression is not on the same line as the block start.
  do |config_home_param: :config_home, ...
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/tmp/3359.rb:2:39: C: Style/SpaceAroundBlockParameters: Extra space before block parameter detected.
  do |config_home_param: :config_home, ...
                                      
/tmp/3359.rb:3:41: C: Style/SpaceAroundBlockParameters: Extra space before block parameter detected.
      service_name_param: :service_name, ...
```
and I agree that we should fix this in the `Style/MultilineBlockLayout` and `Style/SpaceAroundBlockParameters` cops, so they accept the code as I laid it out here.

**Comment by bbatsov** [comment]
I'm closing this ticket due to no recent activity. Feel free to re-open it if you ever come back to it.

**Comment by AlexWayfer** [comment]
I still don't know how to resolve situations with 5 long-named arguments with defaults, except huge multiple `options.fetch`.

**Comment by mikegee** [comment]
Perhaps those 5 arguments (or a subset) can be extracted into an object?

**Comment by AlexWayfer** [comment]
> Perhaps those 5 arguments (or a subset) can be extracted into an object?

A new class for one method (block) and two calls? I don't think this is good decision.

**Comment by AlexWayfer** [comment]
> I'm closing this ticket due to no recent activity.

What kind of activity are you expecting here?

> Feel free to re-open it if you ever come back to it.

I can't, for example.

**Comment by mikegee** [comment]
> What kind of activity are you expecting here?

@jonas054 proposed some changes, but nobody took it up yet.
